File: leetcode/editor/cn/169_多数元素_test.java
//给定一个大小为 n 的数组 nums ，返回其中的多数元素。多数元素是指在数组中出现次数 大于 ⌊ n/2 ⌋ 的元素。 
//
// 你可以假设数组是非空的，并且给定的数组总是存在多数元素。 
//
// 
//
// 示例 1： 
//
// 
//输入：nums = [3,2,3]
//输出：3 
//
// 示例 2： 
//
// 
//输入：nums = [2,2,1,1,1,2,2]
//输出：2
// 
//
// 
//提示：
//
// 
// n == nums.length 
// 1 <= n <= 5 * 10⁴ 
// -10⁹ <= nums[i] <= 10⁹ 
// 
//
// 
//
// 进阶：尝试设计时间复杂度为 O(n)、空间复杂度为 O(1) 的算法解决此问题。 
//
// Related Topics 数组 哈希表 分治 计数 排序 
// 👍 1847 👎 0


import org.junit.jupiter.api.Test;

import static org.assertj.core.api.BDDAssertions.then;


class MajorityElementTest {

    final Solution solution = new Solution();

    @Test
    void givenNormal_when_thenSuccess() {
        int ans = solution.majorityElement(new int[]{1, 2, 3, 2, 2, 3, 2});
        then(ans).isEqualTo(2);
    }

    //leetcode submit region begin(Prohibit modification and deletion)
    class Solution {

        public int majorityElement(int[] nums) {
            int ans = 0, cnt = 0;
            for (int n : nums) {
                if (cnt == 0) {
                    ans = n;
                    cnt += 1;
                } else if (n == ans) {
                    cnt += 1;
                } else {
                    cnt -= 1;
                }
            }
            return ans;
        }

    }
//leetcode submit region end(Prohibit modification and deletion)


}

File: leetcode/editor/cn/169_多数元素_test.py
import unittest
from typing import List


class TestSolution(unittest.TestCase):

    def setUp(self) -> None:
        self.solution = Solution()

    def test_givenNormal_when_thenSuccess(self):
        ans = self.solution.majorityElement([2, 1, 2, 3, 2, 3, 2])
        self.assertEqual(2, ans)


# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        ans, cnt = 0, 0
        for n in nums:
            if cnt == 0:
                ans = n
                cnt += 1
            elif n == ans:
                cnt += 1
            else:
                cnt -= 1
        return ans


# leetcode submit region end(Prohibit modification and deletion)


if __name__ == '__main__':
    unittest.main()
